File: src/main/java/com/supplychain/core/Blockchain.java
package com.supplychain.core;

import java.time.Instant;
import java.util.*;

/**
 * Phase 3: Blockchain Ledger Implementation
 * ==========================================
 * This class implements the blockchain data structure that chains blocks
 * together using cryptographic hashes, creating an immutable ledger.
 *
 * STRUCTURE:
 * Each Block contains:
 * - Index: Position in the chain
 * - Timestamp: When the block was created
 * - Transactions: List of supply chain transactions
 * - Merkle Root: Hash summarizing all transactions in this block
 * - Previous Hash: Hash of the immediately preceding block
 * - Nonce: Proof of work counter (optional, for mining)
 * - Hash: SHA-256 hash of this entire block
 *
 * The chain is immutable because modifying any block would:
 * 1. Change its hash
 * 2. Break the "Previous Hash" link to the next block
 * 3. Invalidate the entire subsequent chain
 */
public class Blockchain {

    /**
     * A single block in the blockchain.
     *
     * Contains all supply chain transactions for a time period and
     * cryptographic links to the previous and next blocks.
     */
    public static class Block {
        public int index;
        public String timestamp;
        public List<Map<String, Object>> transactions;
        public String merkleRoot;
        public String previousHash;
        public int nonce;
        public String hash;

        public Block(int index, String timestamp, List<Map<String, Object>> transactions,
                     String merkleRoot, String previousHash, int nonce) {
            this.index = index;
            this.timestamp = timestamp;
            this.transactions = transactions;
            this.merkleRoot = merkleRoot;
            this.previousHash = previousHash;
            this.nonce = nonce;
            this.hash = calculateHash();
        }

        /**
         * Calculate SHA-256 hash of this block.
         *
         * Hash includes all critical data to ensure tamper evidence.
         * Any change to block contents will produce different hash.
         */
        public String calculateHash() {
            String blockData = index +
                    timestamp +
                    merkleRoot +
                    previousHash +
                    nonce;
            return HashUtils.calculateSHA256(blockData);
        }

        /**
         * Serialize block to map.
         */
        public Map<String, Object> toMap() {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("index", index);
            map.put("timestamp", timestamp);
            map.put("transactions", transactions);
            map.put("merkleRoot", merkleRoot);
            map.put("previousHash", previousHash);
            map.put("nonce", nonce);
            map.put("hash", hash);
            return map;
        }

        @Override
        public String toString() {
            return String.format("Block{#%d, txs=%d, hash=%s...}",
                    index, transactions.size(),
                    hash.length() > 16 ? hash.substring(0, 16) : hash);
        }
    }

    private static final String GENESIS_PREVIOUS_HASH = "0".repeat(64);  // Hardcoded start for genesis block

    private List<Block> chain;
    private List<Map<String, Object>> pendingTransactions;

    /**
     * Initialize empty blockchain.
     */
    public Blockchain() {
        this.chain = new ArrayList<>();
        this.pendingTransactions = new ArrayList<>();
        createGenesisBlock();
    }

    /**
     * Create the first block in the chain (Genesis Block).
     *
     * The genesis block has:
     * - Index 0
     * - Previous hash of all zeros (conventional starting point)
     * - Minimal transaction (system initialization data)
     *
     * @return The genesis block that was added to the chain
     */
    public Block createGenesisBlock() {
        Map<String, Object> genesisTx = HashUtils.createTransaction(
                "GENESIS",
                "SYSTEM",
                "NETWORK",
                "Block 0",
                createMetadata("genesis_block", "Supply Chain Network Genesis")
        );

        // Hash the transaction for Merkle tree
        List<String> txHashes = new ArrayList<>();
        txHashes.add(HashUtils.hashTransaction(genesisTx));

        MerkleTree merkle = new MerkleTree(txHashes);

        Block genesisBlock = new Block(
                0,
                Instant.now().toString(),
                new ArrayList<>(Collections.singletonList(genesisTx)),
                merkle.getMerkleRoot(),
                GENESIS_PREVIOUS_HASH,
                0
        );

        chain.add(genesisBlock);
        return genesisBlock;
    }

    /**
     * Create metadata map with type and description.
     */
    private Map<String, Object> createMetadata(String type, String description) {
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("type", type);
        metadata.put("description", description);
        return metadata;
    }

    /**
     * Get the most recent block in the chain.
     */
    public Block getLatestBlock() {
        return chain.get(chain.size() - 1);
    }

    /**
     * Add a new block to the chain containing the given transactions.
     *
     * This is the core operation of the blockchain:
     * 1. Create Merkle tree from transactions
     * 2. Link to previous block
     * 3. Calculate cryptographic hash
     * 4. Add to chain
     *
     * @param transactions List of transaction maps to include
     * @return The newly created and added block
     */
    public Block addBlock(List<Map<String, Object>> transactions) {
        if (transactions == null) {
            transactions = new ArrayList<>(pendingTransactions);
            pendingTransactions.clear();
        }

        // Hash transactions for Merkle tree
        List<String> txHashes = new ArrayList<>();
        for (Map<String, Object> tx : transactions) {
            txHashes.add(HashUtils.hashTransaction(tx));
        }

        MerkleTree merkleTree = new MerkleTree(txHashes);
        String merkleRoot = merkleTree.getMerkleRoot();

        // Create new block linked to previous
        Block newBlock = new Block(
                chain.size(),
                Instant.now().toString(),
                transactions,
                merkleRoot,
                getLatestBlock().hash,
                0
        );

        chain.add(newBlock);
        return newBlock;
    }

    /**
     * Add a transaction to the pending pool.
     *
     * Pending transactions are waiting to be mined (added to a block).
     *
     * @param transaction Transaction map to add to pending pool
     */
    public void addPendingTransaction(Map<String, Object> transaction) {
        pendingTransactions.add(transaction);
    }

    /**
     * Mine all pending transactions into a new block.
     *
     * @return The newly created block with pending transactions
     */
    public Block minePendingTransactions() {
        return addBlock(null);
    }

    /**
     * Validate the entire blockchain integrity.
     *
     * Checks performed:
     * 1. Each block's hash is correctly calculated
     * 2. Each block's "previous hash" matches the actual previous block's hash
     * 3. Merkle root matches the transactions in each block
     *
     * @return true if chain is valid and tamper-free, false if compromised
     *
     * COMPLEXITY: O(n) where n is number of blocks
     */
    public boolean isValid() {
        for (int i = 1; i < chain.size(); i++) {
            Block currentBlock = chain.get(i);
            Block previousBlock = chain.get(i - 1);

            // Check 1: Hash integrity
            if (!currentBlock.hash.equals(currentBlock.calculateHash())) {
                System.out.println("✗ Block " + i + " hash mismatch!");
                return false;
            }

            // Check 2: Chain integrity (previous hash link)
            if (!currentBlock.previousHash.equals(previousBlock.hash)) {
                System.out.println("✗ Block " + i + " previous hash doesn't match block " + (i-1) + "!");
                return false;
            }

            // Check 3: Merkle root validity
            List<String> txHashes = new ArrayList<>();
            for (Map<String, Object> tx : currentBlock.transactions) {
                txHashes.add(HashUtils.hashTransaction(tx));
            }
            MerkleTree merkle = new MerkleTree(txHashes);
            if (!merkle.getMerkleRoot().equals(currentBlock.merkleRoot)) {
                System.out.println("✗ Block " + i + " Merkle root invalid!");
                return false;
            }
        }

        return true;
    }

    /**
     * Retrieve a specific block by index.
     *
     * @param index Block number to retrieve
     * @return Block if found, null if index out of range
     */
    public Block getBlock(int index) {
        if (index >= 0 && index < chain.size()) {
            return chain.get(index);
        }
        return null;
    }

    /**
     * Trace the complete journey of a product through the supply chain.
     *
     * Searches all blocks and transactions to find every occurrence
     * of the specified productId, returning them in chronological order.
     *
     * @param productId The product identifier to trace
     * @return List of transaction info maps involving this product
     */
    public List<Map<String, Object>> getProductHistory(String productId) {
        List<Map<String, Object>> history = new ArrayList<>();

        for (Block block : chain) {
            for (Map<String, Object> tx : block.transactions) {
                if (productId.equals(tx.get("product_id"))) {
                    Map<String, Object> record = new LinkedHashMap<>();
                    record.put("block_index", block.index);
                    record.put("block_hash", block.hash.substring(0, 16) + "...");
                    record.put("timestamp", tx.get("timestamp"));
                    record.put("sender", tx.get("sender"));
                    record.put("receiver", tx.get("receiver"));
                    record.put("location", tx.get("location"));
                    record.put("verified", true);  // Proven by blockchain
                    history.add(record);
                }
            }
        }

        return history;
    }

    /**
     * Verify product authenticity by checking its complete history.
     *
     * This is the end-to-end verification that consumers use when
     * scanning a product QR code.
     *
     * @param productId Product to verify
     * @return Map with verification results
     */
    public Map<String, Object> verifyProduct(String productId) {
        List<Map<String, Object>> history = getProductHistory(productId);

        if (history.isEmpty()) {
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("verified", false);
            result.put("reason", "Product not found in blockchain");
            result.put("authentic", false);
            return result;
        }

        // Check if product originated from valid genesis-linked manufacturer
        Map<String, Object> origin = history.get(0);
        if ("SYSTEM".equals(origin.get("sender"))) {
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("verified", false);
            result.put("reason", "Invalid origin (genesis block only)");
            result.put("authentic", false);
            return result;
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("verified", true);
        result.put("authentic", true);
        result.put("origin", origin);
        result.put("current_location", history.get(history.size() - 1).get("location"));
        result.put("current_owner", history.get(history.size() - 1).get("receiver"));
        result.put("journey_length", history.size());
        result.put("journey", history);

        return result;
    }

    /**
     * Get the total number of blocks in the chain.
     */
    public int getChainLength() {
        return chain.size();
    }

    /**
     * Get pending transaction count.
     */
    public int getPendingCount() {
        return pendingTransactions.size();
    }

    // =========================================================================
    // UNIT TESTS FOR BLOCKCHAIN
    // =========================================================================

    public static void testGenesisBlock() {
        Blockchain chain = new Blockchain();

        assert chain.getChainLength() == 1 : "Should start with genesis block";
        assert chain.getBlock(0).index == 0 : "Genesis block index should be 0";
        assert chain.getBlock(0).previousHash.equals(GENESIS_PREVIOUS_HASH) : "Genesis previous hash should be zeros";
        assert !chain.getBlock(0).hash.isEmpty() : "Genesis block should have a hash";

        System.out.println("✓ Genesis block created correctly");
    }

    public static void testBlockChaining() {
        Blockchain chain = new Blockchain();

        Map<String, Object> tx1 = HashUtils.createTransaction("PROD-001", "Factory", "Distributor", "Shanghai", null);
        chain.addBlock(Collections.singletonList(tx1));

        Map<String, Object> tx2 = HashUtils.createTransaction("PROD-001", "Distributor", "Retailer", "Beijing", null);
        chain.addBlock(Collections.singletonList(tx2));

        assert chain.getChainLength() == 3 : "Should have 3 blocks (genesis + 2)";
        assert chain.getBlock(2).previousHash.equals(chain.getBlock(1).hash) : "Chain link broken";

        System.out.println("✓ Blocks properly chained");
    }

    public static void testChainValidation() {
        Blockchain chain = new Blockchain();

        // Add some blocks
        for (int i = 0; i < 5; i++) {
            Map<String, Object> tx = HashUtils.createTransaction(
                    String.format("PROD-%03d", i),
                    String.format("Factory %d", i),
                    String.format("Dist %d", i),
                    String.format("Loc %d", i),
                    null
            );
            chain.addBlock(Collections.singletonList(tx));
        }

        assert chain.isValid() : "Valid chain should pass validation";

        // Tamper with a block
        List<Map<String, Object>> tamperedTxs = new ArrayList<>(chain.getBlock(2).transactions);
        Map<String, Object> tamperedTx = new LinkedHashMap<>(tamperedTxs.get(0));
        tamperedTx.put("location", "Tampered Location");
        tamperedTxs.set(0, tamperedTx);

        // Recreate block with tampered data (simulating tampering)
        Block tamperedBlock = new Block(
                2,
                chain.getBlock(2).timestamp,
                tamperedTxs,
                chain.getBlock(2).merkleRoot,
                chain.getBlock(2).previousHash,
                chain.getBlock(2).nonce
        );
        chain.chain.set(2, tamperedBlock);

        assert !chain.isValid() : "Tampered chain should fail validation";

        System.out.println("✓ Chain validation detects tampering");
    }

    public static void testProductHistory() {
        Blockchain chain = new Blockchain();

        // Simulate product journey
        List<String[]> journey = Arrays.asList(
                new String[]{"Factory-A", "Distributor-B", "Shanghai"},
                new String[]{"Distributor-B", "Wholesaler-C", "Beijing"},
                new String[]{"Wholesaler-C", "Retailer-D", "Shanghai"},
                new String[]{"Retailer-D", "Consumer", "Customer Home"}
        );

        for (String[] step : journey) {
            Map<String, Object> tx = HashUtils.createTransaction(
                    "AUTHENTIC-001",
                    step[0],
                    step[1],
                    step[2],
                    null
            );
            chain.addBlock(Collections.singletonList(tx));
        }

        List<Map<String, Object>> history = chain.getProductHistory("AUTHENTIC-001");

        assert history.size() == 4 : "Should have 4 transactions, got " + history.size();
        assert "Factory-A".equals(history.get(0).get("sender")) : "Origin should be factory";
        assert "Consumer".equals(history.get(history.size() - 1).get("receiver")) : "Final destination should be consumer";

        System.out.println("✓ Product journey tracked: " + history.size() + " steps");
        for (Map<String, Object> step : history) {
            System.out.println("  " + step.get("sender") + " → " + step.get("receiver") + " (" + step.get("location") + ")");
        }
    }

    public static void testVerifyProduct() {
        Blockchain chain = new Blockchain();

        // Create authentic product
        Map<String, Object> tx1 = HashUtils.createTransaction("PROD-REAL", "Factory", "Dist", "Factory Location", null);
        chain.addBlock(Collections.singletonList(tx1));

        Map<String, Object> tx2 = HashUtils.createTransaction("PROD-REAL", "Dist", "Retailer", "Distribution Center", null);
        chain.addBlock(Collections.singletonList(tx2));

        // Verify authentic product
        Map<String, Object> result = chain.verifyProduct("PROD-REAL");
        assert (Boolean) result.get("verified") : "Real product should verify";
        assert (Boolean) result.get("authentic") : "Real product should be authentic";

        // Try to verify non-existent product
        result = chain.verifyProduct("PROD-FAKE");
        assert !(Boolean) result.get("verified") : "Fake product should not verify";
        assert !(Boolean) result.get("authentic") : "Fake product should not be authentic";

        System.out.println("✓ Product verification working correctly");
    }

    public static void testMerkleRootInBlock() {
        Blockchain chain = new Blockchain();

        // Add block with multiple transactions
        List<Map<String, Object>> transactions = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            transactions.add(HashUtils.createTransaction(
                    String.format("PROD-%03d", i),
                    String.format("Sender %d", i),
                    String.format("Receiver %d", i),
                    String.format("Loc %d", i),
                    null
            ));
        }

        Block block = chain.addBlock(transactions);

        // Verify Merkle root matches (need to hash transactions)
        List<String> txHashes = new ArrayList<>();
        for (Map<String, Object> tx : transactions) {
            txHashes.add(HashUtils.hashTransaction(tx));
        }
        MerkleTree merkle = new MerkleTree(txHashes);

        assert block.merkleRoot.equals(merkle.getMerkleRoot()) : "Merkle root mismatch";

        System.out.println("✓ Merkle root correctly stored in block");
    }

    /**
     * Run all Blockchain tests.
     */
    public static void runAllTests() {
        System.out.println("============================================================");
        System.out.println("PHASE 3: BLOCKCHAIN LEDGER UNIT TESTS");
        System.out.println("============================================================\n");

        testGenesisBlock();
        System.out.println();
        testBlockChaining();
        System.out.println();
        testChainValidation();
        System.out.println();
        testProductHistory();
        System.out.println();
        testVerifyProduct();
        System.out.println();
        testMerkleRootInBlock();

        System.out.println("\n============================================================");
        System.out.println("ALL TESTS PASSED ✓");
        System.out.println("============================================================\n");
    }

    public static void main(String[] args) {
        runAllTests();
    }
}

File: src/core/blockchain.py
"""
Phase 3: Blockchain Ledger Implementation
===========================================
This module implements the blockchain data structure that chains blocks
together using cryptographic hashes, creating an immutable ledger.

STRUCTURE:
Each Block contains:
- Index: Position in the chain
- Timestamp: When the block was created
- Transactions: List of supply chain transactions
- Merkle Root: Hash summarizing all transactions in this block
- Previous Hash: Hash of the immediately preceding block
- Nonce: Proof of work counter (optional, for mining)
- Hash: SHA-256 hash of this entire block

The chain is immutable because modifying any block would:
1. Change its hash
2. Break the "Previous Hash" link to the next block
3. Invalidate the entire subsequent chain

Author: Blockchain Supply Chain Team
"""

import time
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime

try:
    from .hash_utils import calculate_sha256, hash_transaction, create_transaction
    from .merkle_tree import MerkleTree
except ImportError:
    from hash_utils import calculate_sha256, hash_transaction, create_transaction
    from merkle_tree import MerkleTree


@dataclass
class Block:
    """
    A single block in the blockchain.
    
    Contains all supply chain transactions for a time period and
    cryptographic links to the previous and next blocks.
    """
    index: int
    timestamp: str
    transactions: List[Dict[str, Any]]
    merkle_root: str
    previous_hash: str
    nonce: int = 0
    hash: str = ""
    
    def __post_init__(self):
        """Calculate hash after initialization if not provided."""
        if not self.hash:
            self.hash = self.calculate_hash()
    
    def calculate_hash(self) -> str:
        """
        Calculate SHA-256 hash of this block.
        
        Hash includes all critical data to ensure tamper evidence.
        Any change to block contents will produce different hash.
        """
        block_data = (
            str(self.index) +
            self.timestamp +
            self.merkle_root +
            self.previous_hash +
            str(self.nonce)
        )
        return calculate_sha256(block_data)
    
    def to_dict(self) -> Dict[str, Any]:
        """Serialize block to dictionary."""
        return {
            'index': self.index,
            'timestamp': self.timestamp,
            'transactions': self.transactions,
            'merkle_root': self.merkle_root,
            'previous_hash': self.previous_hash,
            'nonce': self.nonce,
            'hash': self.hash
        }


class Blockchain:
    """
    The complete blockchain ledger for supply chain tracking.
    
    KEY OPERATIONS:
    1. create_genesis_block(): Initialize the first block
    2. add_block(transactions): Add new block with transactions
    3. is_valid(): Verify entire chain integrity
    4. get_block(index): Retrieve specific block
    5. get_transaction_history(product_id): Trace product journey
    
    USAGE:
        >>> chain = Blockchain()
        >>> chain.add_block([tx1, tx2, tx3])
        >>> chain.is_valid()
        True
        >>> chain.get_product_history('PROD-001')
        [{'from': 'Factory', 'to': 'Dist'}, ...]
    """
    
    GENESIS_PREVIOUS_HASH = "0" * 64  # Hardcoded start for genesis block
    
    def __init__(self):
        """Initialize empty blockchain."""
        self.chain: List[Block] = []
        self.pending_transactions: List[Dict[str, Any]] = []
        
        # Create genesis block
        self.create_genesis_block()
    
    def create_genesis_block(self) -> Block:
        """
        Create the first block in the chain (Genesis Block).
        
        The genesis block has:
        - Index 0
        - Previous hash of all zeros (conventional starting point)
        - Empty or minimal transactions (often includes system initialization data)
        
        Returns:
            The genesis block that was added to the chain
        """
        genesis_tx = create_transaction(
            product_id='GENESIS',
            sender='SYSTEM',
            receiver='NETWORK',
            location='Block 0',
            metadata={'type': 'genesis_block', 'description': 'Supply Chain Network Genesis'}
        )
        
        # Hash the transaction for Merkle tree
        tx_hashes = [hash_transaction(genesis_tx)]
        merkle = MerkleTree(tx_hashes)
        
        genesis_block = Block(
            index=0,
            timestamp=datetime.utcnow().isoformat() + 'Z',
            transactions=[genesis_tx],
            merkle_root=merkle.get_merkle_root(),
            previous_hash=self.GENESIS_PREVIOUS_HASH,
            nonce=0
        )
        
        self.chain.append(genesis_block)
        return genesis_block
    
    def get_latest_block(self) -> Block:
        """Get the most recent block in the chain."""
        return self.chain[-1]
    
    def add_block(self, transactions: List[Dict[str, Any]] = None) -> Block:
        """
        Add a new block to the chain containing the given transactions.
        
        This is the core operation of the blockchain:
        1. Create Merkle tree from transactions
        2. Link to previous block
        3. Calculate cryptographic hash
        4. Add to chain
        
        Args:
            transactions: List of transaction dictionaries to include
            
        Returns:
            The newly created and added block
        """
        if transactions is None:
            transactions = self.pending_transactions.copy()
            self.pending_transactions.clear()
        
        # Hash transactions for Merkle tree
        tx_hashes = [hash_transaction(tx) for tx in transactions]
        merkle_tree = MerkleTree(tx_hashes)
        merkle_root = merkle_tree.get_merkle_root()
        
        # Create new block linked to previous
        new_block = Block(
            index=len(self.chain),
            timestamp=datetime.utcnow().isoformat() + 'Z',
            transactions=transactions,
            merkle_root=merkle_root,
            previous_hash=self.get_latest_block().hash,
            nonce=0
        )
        
        self.chain.append(new_block)
        return new_block
    
    def add_pending_transaction(self, transaction: Dict[str, Any]) -> None:
        """
        Add a transaction to the pending pool.
        
        Pending transactions are waiting to be mined (added to a block).
        This simulates real-world blockchain where transactions wait
        until a miner includes them in a block.
        
        Args:
            transaction: Transaction dictionary to add to pending pool
        """
        self.pending_transactions.append(transaction)
    
    def mine_pending_transactions(self) -> Block:
        """
        Mine all pending transactions into a new block.
        
        In a real blockchain, this would involve proof-of-work.
        For our supply chain, we use a simplified model where
        pending transactions are batched into blocks.
        
        Returns:
            The newly created block with pending transactions
        """
        return self.add_block()
    
    def is_valid(self) -> bool:
        """
        Validate the entire blockchain integrity.
        
        Checks performed:
        1. Each block's hash is correctly calculated
        2. Each block's "previous hash" matches the actual previous block's hash
        3. Merkle root matches the transactions in each block
        
        Returns:
            True if chain is valid and tamper-free, False if compromised
            
        COMPLEXITY: O(n) where n is number of blocks
        """
        for i in range(1, len(self.chain)):
            current_block = self.chain[i]
            previous_block = self.chain[i - 1]
            
            # Check 1: Hash integrity
            if current_block.hash != current_block.calculate_hash():
                print(f"✗ Block {i} hash mismatch!")
                return False
            
            # Check 2: Chain integrity (previous hash link)
            if current_block.previous_hash != previous_block.hash:
                print(f"✗ Block {i} previous hash doesn't match block {i-1}!")
                return False
            
            # Check 3: Merkle root validity
            tx_hashes = [hash_transaction(tx) for tx in current_block.transactions]
            merkle = MerkleTree(tx_hashes)
            if merkle.get_merkle_root() != current_block.merkle_root:
                print(f"✗ Block {i} Merkle root invalid!")
                return False
        
        return True
    
    def get_block(self, index: int) -> Optional[Block]:
        """
        Retrieve a specific block by index.
        
        Args:
            index: Block number to retrieve
            
        Returns:
            Block if found, None if index out of range
        """
        if 0 <= index < len(self.chain):
            return self.chain[index]
        return None
    
    def get_product_history(self, product_id: str) -> List[Dict[str, Any]]:
        """
        Trace the complete journey of a product through the supply chain.
        
        Searches all blocks and transactions to find every occurrence
        of the specified product_id, returning them in chronological order.
        
        Args:
            product_id: The product identifier to trace
            
        Returns:
            List of transaction dictionaries involving this product
        """
        history = []
        
        for block in self.chain:
            for tx in block.transactions:
                if tx.get('product_id') == product_id:
                    history.append({
                        'block_index': block.index,
                        'block_hash': block.hash[:16] + '...',
                        'timestamp': tx['timestamp'],
                        'sender': tx['sender'],
                        'receiver': tx['receiver'],
                        'location': tx['location'],
                        'verified': True  # Proven by blockchain
                    })
        
        return history
    
    def verify_product(self, product_id: str, expected_hash: str = None) -> Dict[str, Any]:
        """
        Verify product authenticity by checking its complete history.
        
        This is the end-to-end verification that consumers use when
        scanning a product QR code.
        
        Args:
            product_id: Product to verify
            expected_hash: Optional hash of a specific transaction to verify
            
        Returns:
            Dictionary with verification results
        """
        history = self.get_product_history(product_id)
        
        if not history:
            return {
                'verified': False,
                'reason': 'Product not found in blockchain',
                'authentic': False
            }
        
        # Check if product originated from valid genesis-linked manufacturer
        origin = history[0]
        if origin['sender'] == 'SYSTEM':
            return {
                'verified': False,
                'reason': 'Invalid origin (genesis block only)',
                'authentic': False
            }
        
        result = {
            'verified': True,
            'authentic': True,
            'origin': origin,
            'current_location': history[-1]['location'],
            'current_owner': history[-1]['receiver'],
            'journey_length': len(history),
            'journey': history
        }
        
        return result
    
    def get_chain_length(self) -> int:
        """Get the total number of blocks in the chain."""
        return len(self.chain)
    
    def display_chain(self) -> str:
        """
        Generate ASCII representation of the blockchain.
        
        Returns:
            String showing chain structure with key information
        """
        lines = []
        lines.append("\n" + "=" * 70)
        lines.append("BLOCKCHAIN LEDGER")
        lines.append("=" * 70)
        
        for block in self.chain:
            lines.append(f"\n┌─ Block #{block.index}")
            lines.append(f"│  Timestamp: {block.timestamp}")
            lines.append(f"│  Transactions: {len(block.transactions)}")
            lines.append(f"│  Merkle Root: {block.merkle_root[:32]}...")
            lines.append(f"│  Previous Hash: {block.previous_hash[:32]}...")
            lines.append(f"│  Block Hash: {block.hash[:32]}...")
            lines.append(f"│  Status: ✓ Verified")
            lines.append("└" + "─" * 68)
        
        lines.append(f"\nTotal Blocks: {len(self.chain)}")
        lines.append(f"Pending Transactions: {len(self.pending_transactions)}")
        lines.append("=" * 70 + "\n")
        
        return "\n".join(lines)


# ============================================================================
# UNIT TESTS FOR BLOCKCHAIN
# ============================================================================

def test_genesis_block():
    """Test that genesis block is created correctly."""
    chain = Blockchain()
    
    assert len(chain.chain) == 1, "Should start with genesis block"
    assert chain.chain[0].index == 0, "Genesis block index should be 0"
    assert chain.chain[0].previous_hash == "0" * 64, "Genesis previous hash should be zeros"
    assert chain.chain[0].hash != "", "Genesis block should have a hash"
    
    print("✓ Genesis block created correctly")


def test_block_chaining():
    """Test that blocks are properly chained together."""
    chain = Blockchain()
    
    tx1 = create_transaction('PROD-001', 'Factory', 'Distributor', 'Shanghai')
    chain.add_block([tx1])
    
    tx2 = create_transaction('PROD-001', 'Distributor', 'Retailer', 'Beijing')
    chain.add_block([tx2])
    
    assert len(chain.chain) == 3, "Should have 3 blocks (genesis + 2)"
    assert chain.chain[2].previous_hash == chain.chain[1].hash, "Chain link broken"
    
    print("✓ Blocks properly chained")


def test_chain_validation():
    """Test blockchain integrity validation."""
    chain = Blockchain()
    
    # Add some blocks
    for i in range(5):
        tx = create_transaction(f'PROD-{i:03d}', f'Factory {i}', f'Dist {i}', f'Loc {i}')
        chain.add_block([tx])
    
    assert chain.is_valid() == True, "Valid chain should pass validation"
    
    # Tamper with a block
    chain.chain[2].transactions[0]['location'] = 'Tampered Location'
    
    assert chain.is_valid() == False, "Tampered chain should fail validation"
    
    print("✓ Chain validation detects tampering")


def test_product_history():
    """Test tracing product journey through supply chain."""
    chain = Blockchain()
    
    # Simulate product journey
    journey = [
        ('Factory-A', 'Distributor-B', 'Shanghai'),
        ('Distributor-B', 'Wholesaler-C', 'Beijing'),
        ('Wholesaler-C', 'Retailer-D', 'Shanghai'),
        ('Retailer-D', 'Consumer', 'Customer Home'),
    ]
    
    for sender, receiver, location in journey:
        tx = create_transaction('AUTHENTIC-001', sender, receiver, location)
        chain.add_block([tx])
    
    history = chain.get_product_history('AUTHENTIC-001')
    
    assert len(history) == 4, f"Should have 4 transactions, got {len(history)}"
    assert history[0]['sender'] == 'Factory-A', "Origin should be factory"
    assert history[-1]['receiver'] == 'Consumer', "Final destination should be consumer"
    
    print(f"✓ Product journey tracked: {len(history)} steps")
    for step in history:
        print(f"  {step['sender']} → {step['receiver']} ({step['location']})")


def test_verify_product():
    """Test product verification functionality."""
    chain = Blockchain()
    
    # Create authentic product
    tx1 = create_transaction('PROD-REAL', 'Factory', 'Dist', 'Factory Location')
    chain.add_block([tx1])
    
    tx2 = create_transaction('PROD-REAL', 'Dist', 'Retailer', 'Distribution Center')
    chain.add_block([tx2])
    
    # Verify authentic product
    result = chain.verify_product('PROD-REAL')
    assert result['verified'] == True, "Real product should verify"
    assert result['authentic'] == True, "Real product should be authentic"
    
    # Try to verify non-existent product
    result = chain.verify_product('PROD-FAKE')
    assert result['verified'] == False, "Fake product should not verify"
    assert result['authentic'] == False, "Fake product should not be authentic"
    
    print("✓ Product verification working correctly")


def test_merkle_root_in_block():
    """Test that Merkle root is correctly calculated and stored."""
    chain = Blockchain()
    
    # Add block with multiple transactions
    transactions = [
        create_transaction(f'PROD-{i:03d}', f'Sender {i}', f'Receiver {i}', f'Loc {i}')
        for i in range(4)
    ]
    
    block = chain.add_block(transactions)
    
    # Verify Merkle root matches (need to hash transactions)
    tx_hashes = [hash_transaction(tx) for tx in transactions]
    merkle = MerkleTree(tx_hashes)
    assert block.merkle_root == merkle.get_merkle_root(), "Merkle root mismatch"
    
    print("✓ Merkle root correctly stored in block")


if __name__ == '__main__':
    print("=" * 60)
    print("PHASE 3: BLOCKCHAIN LEDGER UNIT TESTS")
    print("=" * 60)
    
    test_genesis_block()
    print()
    test_block_chaining()
    print()
    test_chain_validation()
    print()
    test_product_history()
    print()
    test_verify_product()
    print()
    test_merkle_root_in_block()
    
    print("\n" + "=" * 60)
    print("ALL TESTS PASSED ✓")
    print("=" * 60)
    print("\nBlockchain ledger complete. Proceeding to supply chain logic...")
